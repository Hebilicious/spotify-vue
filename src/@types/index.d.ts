export type Track = {
    _id: string,
    name: string,
    year: number | null,
    artist: string | null,
    album: string | null,
    duration_s: number,
    imageUri: string,
    type: string,
    uri: string
};

export type Album = {
    type: string,
    _id: string,
    name: string,
    year: number,
    total_tracks: number,
    description: string,
    imageUri: string,
    tracks: Track[]
};

export type PlayList = Album[]