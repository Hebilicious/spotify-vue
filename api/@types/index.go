package api

type Track struct {
	ID       string `json:"_id"`
	Name     string `json:"name"`
	Year     int    `json:"year"`
	Artist   string `json:"artist"`
	Album    string `json:"album"`
	Duration int    `json:"duration_s"`
	ImageUri string `json:"imageUri"`
	Type     string `json:"type"`
	Uri      string `json:"uri"`
}

type Album struct {
	ID          string  `json:"_id"`
	Type        string  `json:"type"`
	Name        string  `json:"name"`
	Year        int     `json:"year"`
	TotalTracks int     `json:"total_tracks"`
	Description string  `json:"description"`
	ImageUri    string  `json:"imageUri"`
	Tracks      []Track `json:"tracks"`
}

type PlayList []Album
