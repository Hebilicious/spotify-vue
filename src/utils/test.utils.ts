import { VueWrapper } from '@vue/test-utils';
import { expect, it } from 'vitest';

export const testSnapShot = (wrapperGetter: () => VueWrapper) => {
  it('Matches the snapshot', () => {
    expect(wrapperGetter().html()).toMatchSnapshot();
  });
};
