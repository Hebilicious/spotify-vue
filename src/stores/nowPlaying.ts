import { defineStore } from 'pinia';

import { Track } from '@/@types';

export const useNowPlayingStore = defineStore({
  id: 'playback-queue',
  state: () => ({
    playQueue: ([] as Track[]),
    currentIndex: (undefined as unknown as number),
  }),
  getters: {
    getCurrentPlaylist: ({ playQueue }) => playQueue,
    getCurrentPlaying: ({ playQueue, currentIndex }) => currentIndex !== undefined ? playQueue?.[currentIndex] : (undefined as unknown as Track),
  },
  actions: {
    setCurrentPlaylist(newPlayQueue: Track[]) {
      this.$patch({ playQueue: newPlayQueue, currentIndex: 0 });
    },
    next() {
      (this.currentIndex < (this.playQueue.length - 1)) && this.currentIndex++;
    },
    prev() {
      this.currentIndex > 0 && this.currentIndex--;
    },
    clearPlaylist() {
      this.playQueue = [];
      this.currentIndex = (undefined as unknown as number);
    }
  },
});
