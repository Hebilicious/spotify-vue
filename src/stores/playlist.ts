import { PlayList } from '@/@types';
import { getAllPlayLists } from '@/services/playListService';
import { defineStore } from 'pinia';

export const usePlayListStore = defineStore({
    id: 'playlists',
    state: () => ({
        playlists: ([] as PlayList),
    }),
    getters: {
        getPlayLists: ({ playlists }) => playlists,
    },
    actions: {
        async loadPlayLists() {
            this.playlists = await getAllPlayLists();
        }
    },
});

export default usePlayListStore;