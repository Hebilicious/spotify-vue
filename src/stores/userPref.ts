import { Track } from '@/@types';
import { defineStore } from 'pinia';
import { useNowPlayingStore } from './nowPlaying'

export const useUserPrefStore = defineStore({
    id: 'user-preferences',
    state: () => ({
        likedSongs: ([] as Track[]),
        volumePref: 0.6
    }),
    getters: {
        getLikedSongs: ({ likedSongs }) => likedSongs,
        isLikedSong: ({ likedSongs }) => {
            const playbackStore = useNowPlayingStore();
            return playbackStore.getCurrentPlaying && likedSongs.find(({ _id }) => _id === playbackStore.getCurrentPlaying._id)
        },
        getVolumePref: ({ volumePref }) => volumePref
    },
    actions: {
        toggleLikedSong(track: Track) {
            this.likedSongs = this.isLikedSong
                ? this.likedSongs.filter(({ _id }) => _id !== track._id)
                : [...this.likedSongs, track]
        },

        setVolumePref(volumePref: number) {
            this.volumePref = volumePref;
        }
    },
});
