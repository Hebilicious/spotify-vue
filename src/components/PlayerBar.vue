<template>
  <div class="player">
    <div class="now-playing-meta">
      <template v-if="nowPlayingStore.getCurrentPlaying">
        <img :src="nowPlayingStore.getCurrentPlaying.imageUri" alt="now-playing-cover-art" />
        <div class="now-playing-info">
          <span class="now-playing-title">{{ nowPlayingStore.getCurrentPlaying?.name }}</span>
          <span class="now-playing-artist">{{ nowPlayingStore.getCurrentPlaying?.artist }}</span>
        </div>
        <span
          :class="`material-icons ${userPrefStore.isLikedSong ? 'selected' : 'unselected'}`"
          @click="toggleFavorite"
          data-testid="favorite"
        >favorite</span>
      </template>
    </div>
    <div class="player-controls">
      <audio
        ref="audioRef"
        :src="nowPlayingStore.getCurrentPlaying?.uri || '#'"
        preload="metadata"
        hidden
      ></audio>
      <div>
        <span
          class="material-icons playback-icon"
          @click="nowPlayingStore.prev"
          data-testid="prev"
        >skip_previous</span>
        <span
          class="material-icons playback-icon"
          @click="play()"
          data-testid="play"
        >{{ state.isPlaying ? 'pause_circle' : 'play_circle' }}</span>
        <span
          class="material-icons playback-icon"
          @click="nowPlayingStore.next"
          data-testid="next"
        >skip_next</span>
      </div>
      <input
        type="range"
        id="seek-slider"
        class="playback-slider"
        :max="audioRef?.duration"
        :value="state.currentPlaybackTime"
        @change="changeCurrentPlayback"
      />
    </div>
    <div class="now-playing-additional">
      <span class="material-icons unselected">queue_music</span>
      <div class="player-volume">
        <span class="material-icons unselected">volume_up</span>
        <input
          type="range"
          id="volume-slider"
          max="100"
          :value="userPrefStore.getVolumePref * 100"
          @change="changeVolume"
          @mouseenter="toggleShowVolume('enter')"
          @mouseleave="toggleShowVolume('exit')"
        />
      </div>
    </div>
  </div>
</template>

<script lang="ts" setup>
import { reactive, watch } from 'vue';
import { useNowPlayingStore } from '@/stores/nowPlaying';
import { useUserPrefStore } from '@/stores/userPref';

const nowPlayingStore = useNowPlayingStore();
const userPrefStore = useUserPrefStore();

interface PlayerState {
  isPlaying: boolean;
  currentPlaybackTime: number;
}

const audioRef = $ref<HTMLAudioElement | null>(null);

const state = reactive<PlayerState>({
  isPlaying: false,
  currentPlaybackTime: 0,
});

const changeCurrentPlayback = (event: Event) => {
  audioRef!.currentTime = parseInt((<HTMLInputElement>event.target).value);
  state.currentPlaybackTime = audioRef!.currentTime
}

const changeVolume = (event: Event) => {
  audioRef!.volume = parseInt((<HTMLInputElement>event.target).value) / 100
  userPrefStore.setVolumePref(audioRef!.volume)
  toggleShowVolume('enter');
}

const toggleShowVolume = (event: string) => {
  document.getElementById('volume-slider')!.style.background = event === 'enter'
    ? 'linear-gradient(to right, #41b883 0%, #41b883 ' + (userPrefStore.getVolumePref * 100) + '%, #fff ' + (userPrefStore.getVolumePref * 100) + '%, white 100%)'
    : '#d3d3d3'
}

const play = () => {
  audioRef?.paused ? audioRef?.play() : audioRef?.pause()
  state.isPlaying = !state.isPlaying;
};

setInterval(() => {
  state.currentPlaybackTime = audioRef?.currentTime || 0
}, 980)

watch(() => nowPlayingStore, () => {
  state.isPlaying = false;
  play();
}, { flush: 'post', deep: true })

watch(() => audioRef, () => (audioRef && (audioRef.volume = userPrefStore.getVolumePref)), { flush: 'post' })

const toggleFavorite = () => userPrefStore.toggleLikedSong(nowPlayingStore.getCurrentPlaying)

</script>

<style lang="scss">
.player {
  grid-area: now-playing-bar;
  height: 100px;
  background-color: var(--vt-c-black);
  border-top: 1px solid var(--vt-c-black-mute);
  z-index: 4;
  display: flex;
  padding: 1rem;
  align-items: center;
  user-select: none;

  .unselected {
    transform: scale(0.75);
  }

  .now-playing-meta {
    display: flex;
    flex: 1;
    justify-content: flex-start;
    align-items: center;

    img {
      height: 56px;
      width: 56px;
    }

    .selected {
      color: var(--color-accent-primary);
      transform: scale(1.1);
    }

    .selected,
    .unselected {
      cursor: pointer;
      transition: color 0.5s ease-out, transform 1s ease-in;
    }

    .now-playing-info {
      display: flex;
      flex-flow: column;
      padding: 0 1rem 0 1rem;

      .now-playing-title {
        color: var(--color-heading);
        font-weight: 600;
      }

      .now-playing-artist {
        font-size: 12px;
        line-height: 0.75rem;
      }
    }
  }

  .player-controls {
    display: flex;
    flex: 1;
    justify-content: center;
    flex-direction: column;
    align-items: center;

    .playback-icon {
      transform: scale(1.5);
      padding: 1rem;
      color: var(--color-heading);
      cursor: pointer;
    }

    .playback-slider {
      width: 100%;
    }
  }

  .now-playing-additional {
    display: flex;
    justify-content: flex-end;
    flex: 1;
    grid-column-gap: 0.5rem;

    .player-volume {
      display: flex;
      align-items: center;

      input[type="range" i] {
        -webkit-appearance: none; /* Override default CSS styles */
        appearance: none;
        outline: none;
        background: #d3d3d3;
        -webkit-appearance: none;
        width: 100%;
        height: 0.25rem;
        -webkit-transition: 0.2s;
        transition: opacity 0.2s;
        border-radius: 5px;

        &::-webkit-slider-thumb {
          -webkit-appearance: none;
          appearance: none;
        }
      }

      input[type="range" i]:hover {
        &::-webkit-slider-thumb {
          -webkit-appearance: none;
          appearance: none;
          height: 0.75rem;
          width: 0.75rem;
          border-radius: 50%;
          background: var(--color-accent-primary);
          cursor: pointer;
        }
      }
    }
  }
}
</style>
