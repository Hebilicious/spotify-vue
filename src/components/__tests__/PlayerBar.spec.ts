import { Track } from "@/@types";
import { useNowPlayingStore } from "@/stores/nowPlaying";
import { useUserPrefStore } from "@/stores/userPref";
import { testSnapShot } from "@/utils/test.utils";
import { createTestingPinia } from "@pinia/testing";
import { shallowMount, VueWrapper } from "@vue/test-utils";
import { VueNode } from "@vue/test-utils/dist/types";
import { afterEach, beforeEach, describe, expect, fn, it } from "vitest";
import PlayerBar from "../PlayerBar.vue";

describe('PlayerBar', () => {
    let wrapper: VueWrapper;

    describe('When no playlist is selected', () => {
        beforeEach(() => {
            wrapper = shallowMount(PlayerBar, {
                global: {
                    plugins: [createTestingPinia({ createSpy: () => fn() })]
                }
            });
        });

        afterEach(() => {
            wrapper.unmount();
        });

        it('does not render the song details', () => {
            expect(wrapper.find('.now-playing-info').exists()).toBeFalsy();
        });

        testSnapShot(() => wrapper);
    });

    describe('When a playlist is selected', () => {
        const mockPlayList: Partial<Track>[] = [
            { _id: "0", name: "mockSong1", artist: "mockArtist1", uri: "mockUri1" }
        ];


        beforeEach(() => {
            const pinia = createTestingPinia({
                createSpy: () => fn(),
                stubActions: false
            });
            useNowPlayingStore().setCurrentPlaylist(mockPlayList as Track[]);
            useNowPlayingStore().next = fn();
            useNowPlayingStore().prev = fn();

            wrapper = shallowMount(PlayerBar, {
                global: {
                    plugins: [pinia]
                },
            })
        })

        afterEach(() => wrapper.unmount());

        it('renders the song details when song info is present', async () => {
            expect(wrapper.find('.now-playing-info').exists()).toBeTruthy();
        });

        it('calls next when next icon is clicked', async () => {
            const store = useNowPlayingStore();
            await wrapper.find('[data-testid="next"]').trigger('click');
            expect(store.next).toBeCalled();
        });

        it('calls prev when prev icon is clicked', async () => {
            const store = useNowPlayingStore();
            await wrapper.find('[data-testid="prev"]').trigger('click');
            expect(store.prev).toBeCalled();
        });

        it('calls toggleFavorite when favorite icon is clicked', async () => {
            const store = useUserPrefStore();
            await wrapper.find('[data-testid="favorite"]').trigger('click');
            expect(store.toggleLikedSong).toBeCalled();
        });

        it('calls play when the play icon is clicked', async () => {
            (wrapper.vm as { [key: string]: any }).audioRef = {
                play: fn(),
                pause: fn()
            }
            await wrapper.find('[data-testid="play"]').trigger('click');
            expect((wrapper.vm as any).state.isPlaying).toBeTruthy();
        })

        it('changes current playback time when playback slider is changed', async () => {
            (wrapper.find('.playback-slider').element as VueNode<HTMLInputElement>).value = '80'
            wrapper.find('.playback-slider').trigger('change');
        })

        it('changes current volume when volume slider is changed', async () => {
            document.getElementById = fn().mockImplementation(() => ({
                style: {
                    background: ''
                }
            }))
            const store = useUserPrefStore();
            (wrapper.find('#volume-slider').element as VueNode<HTMLInputElement>).value = '80'
            wrapper.find('#volume-slider').trigger('change');
            expect(store.setVolumePref).toHaveBeenCalledWith(0.8);
        });
        testSnapShot(() => wrapper);

    });
})

