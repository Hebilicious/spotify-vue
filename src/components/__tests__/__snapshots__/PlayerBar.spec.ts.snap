// Vitest Snapshot v1

exports[`PlayerBar > When a playlist is selected > Matches the snapshot 1`] = `
"<div class=\\"player\\">
  <div class=\\"now-playing-meta\\"><img alt=\\"now-playing-cover-art\\">
    <div class=\\"now-playing-info\\"><span class=\\"now-playing-title\\">mockSong1</span><span class=\\"now-playing-artist\\">mockArtist1</span></div><span class=\\"material-icons unselected\\" data-testid=\\"favorite\\">favorite</span>
  </div>
  <div class=\\"player-controls\\"><audio src=\\"mockUri1\\" preload=\\"metadata\\" hidden=\\"\\"></audio>
    <div><span class=\\"material-icons playback-icon\\" data-testid=\\"prev\\">skip_previous</span><span class=\\"material-icons playback-icon\\" data-testid=\\"play\\">play_circle</span><span class=\\"material-icons playback-icon\\" data-testid=\\"next\\">skip_next</span></div><input type=\\"range\\" id=\\"seek-slider\\" class=\\"playback-slider\\" max=\\"NaN\\">
  </div>
  <div class=\\"now-playing-additional\\"><span class=\\"material-icons unselected\\">queue_music</span>
    <div class=\\"player-volume\\"><span class=\\"material-icons unselected\\">volume_up</span><input type=\\"range\\" id=\\"volume-slider\\" max=\\"100\\"></div>
  </div>
</div>"
`;

exports[`PlayerBar > When no playlist is selected > Matches the snapshot 1`] = `
"<div class=\\"player\\">
  <div class=\\"now-playing-meta\\">
    <!--v-if-->
  </div>
  <div class=\\"player-controls\\"><audio src=\\"#\\" preload=\\"metadata\\" hidden=\\"\\"></audio>
    <div><span class=\\"material-icons playback-icon\\" data-testid=\\"prev\\">skip_previous</span><span class=\\"material-icons playback-icon\\" data-testid=\\"play\\">play_circle</span><span class=\\"material-icons playback-icon\\" data-testid=\\"next\\">skip_next</span></div><input type=\\"range\\" id=\\"seek-slider\\" class=\\"playback-slider\\" max=\\"NaN\\">
  </div>
  <div class=\\"now-playing-additional\\"><span class=\\"material-icons unselected\\">queue_music</span>
    <div class=\\"player-volume\\"><span class=\\"material-icons unselected\\">volume_up</span><input type=\\"range\\" id=\\"volume-slider\\" max=\\"100\\"></div>
  </div>
</div>"
`;
